### ISSUE TITLE
Inlined differs when you change the order of the members, is this correct?

### ISSUE THREAD

**Issue by sonnemaf** [szehetner/InliningAnalyzer#2]
Hi,

I really like your tool. I was looking for that for some while. But I have some strange behavior. Maybe you can explain it.

I have this Employee class in which the call to the Salary (getter) is inlined in the GetYearSalary() method but not in the RaiseSalary() and ToString() methods. The Name (getter) in the ToString() method is also not inlined.

```
internal class Employee {

    public string Name { get; set; }

    private double _salary;

    public Employee(string name, double salary) {
        this.Name = name;
        this.Salary = salary;
    }

    public double GetYearSalary() {
        return this.Salary * 12;
    }

    public double Salary {
        get => _salary;
        set => _salary = value;
    }

    public void RaiseSalary(double percentage) {
        this.Salary += Salary * (percentage / 100);
    }

    public override string ToString() {
        return string.Format("Employee Name = {0}, Salary = {1}", this.Name, this.Salary.ToString());
    }

}
```

I have moved the Salary property and the Name auto-property to the end of the class and then ran the tool again. The Name and Salary (getters and setters) calls in the GetYearSalary(), RaiseSalary() and ToString() are now all inlined. I find it very strange that the order of the members influences the inlining. Is it a bug? In the compiler or in your extension?

```
internal class Employee2 {

    private double _salary;

    public Employee2(string name, double salary) {
        this.Name = name;
        this.Salary = salary;
    }

    public double GetYearSalary() {
        return this.Salary * 12;
    }

    public void RaiseSalary(double percentage) {
        this.Salary += Salary * (percentage / 100);
    }

    public override string ToString() {
        return string.Format("Employee Name = {0}, Salary = {1}", this.Name, this.Salary.ToString());
    }

    public double Salary {
        get => _salary;
        set => _salary = value;
    }

    public string Name { get; set; }

}
```

Regards,

Fons

**Comment by szehetner** [comment]
Thanks, glad you like the tool. That's an interesting case you have there. I can't explain it, but it is not a bug in the extension. The compiled methods really are different.

I checked the ASM via [SharpLab](
https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAIAYAEqCMBuAWACgNscAWQkrXAOgBkBLAOwEcrS6AVAUwA9gHGjgCsVYszBQeEAA5gAxj0wBlKbIA2PYgG9imA5hbAeAJ0kbsAJkwBRKJoD2ATx7aihzHqL7Ph1ADMZFgAclLKOpgA5jyCmBCxeJgAvsS+fpiypowAbmAmmHCOAK4ARlqYAPoQYBpgps7iHhnYQfZOrjwAFLhYktIwhSXlyjV1DQCU6X7eLX7AABaMELRh0pgAvJj9PFRznovLtGrjzpvxtfWN056pPs0ZgUNlFQDisQCaPPUnV11TDxmNxaqAA7JhDitfg1MAAqTA4Kx7Fp3YEGJ5FF7KaHONFePGeGLATYAPiqYyuyP28VipPJlxhWzyGmKuzxqMBnieqHImAASmBljwcV1MSNMmYlMxgGAYgC5rNqRCllCGWdEFscXDMF1ZJKeNLZcoAPQI9DoCZUvwcuZPRw5MzZODKXqYLiOFTAbLMKL/PGK6lg4K0ABijlMUHyXQARO0NC43Jg1sotjp0MlBlrUzhktHBpDVuF8yrjmraO7Pd7fRNLey0oCbYZjGYLNY7A5450rNNvHisrl8soxRVqmqmra2h2EzwrD0cH0i89xRTJv6CYYC8nzjsrXMC1mLqdd4ZG48gsPlO9gF8fmq/ZzDAH9kH92rtYjjwZT35ubyBUKRQvCVTClGU5TXB8MlfU5MA1VQ33hXV9UNGJMFNecLU/FJ6wnTB7UdRhnWCN0PS9FhqwgpVsHBXpQ3DSNgBjONpyTcJzjTDN4Jg7Nc2LI5kz41VTnLUiq3+WtIOw+5cKAnFKKVIk6VHI91wMBJiQ2MllKuc5mVZLDvy5IJXS3SIiSSdSkm/O47iAA)

You can easily see the calls to get_Salary/set_Salary in Employee.RaiseSalary:
```
Sample.Employee.RaiseSalary(Double)
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: sub esp, 0x18
    L0006: xor eax, eax
    L0008: mov [ebp-0x18], eax
    L000b: mov [ebp-0x14], ecx
    L000e: cmp dword [0x13f2d934], 0x0
    L0015: jz L001c
    L0017: call 0x74a20d70
    L001c: nop
    L001d: mov eax, [ebp-0x14]
    L0020: mov [ebp-0x18], eax
    L0023: mov ecx, [ebp-0x14]
    L0026: call Sample.Employee.get_Salary()
    L002b: fstp qword [ebp-0x8]
    L002e: mov ecx, [ebp-0x14]
    L0031: call Sample.Employee.get_Salary()
    L0036: fstp qword [ebp-0x10]
    L0039: fld qword [ebp-0x10]
    L003c: fld qword [ebp+0x8]
    L003f: fdiv dword [0x13fd27c0]
    L0045: fmulp st1, st0
    L0047: fadd qword [ebp-0x8]
    L004a: sub esp, 0x8
    L004d: fstp qword [esp]
    L0050: mov ecx, [ebp-0x18]
    L0053: call Sample.Employee.set_Salary(Double)
    L0058: nop
    L0059: nop
    L005a: mov esp, ebp
    L005c: pop ebp
    L005d: ret 0x8
```

But everything is inlined in Employee2:
```
Sample.Employee2.RaiseSalary(Double)
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: sub esp, 0x18
    L0006: xor eax, eax
    L0008: mov [ebp-0x18], eax
    L000b: mov [ebp-0x14], ecx
    L000e: cmp dword [0x13f2d934], 0x0
    L0015: jz L001c
    L0017: call 0x74a20d70
    L001c: nop
    L001d: mov eax, [ebp-0x14]
    L0020: mov [ebp-0x18], eax
    L0023: mov ecx, [ebp-0x14]
    L0026: call dword [0x13f2db84]
    L002c: fstp qword [ebp-0x8]
    L002f: mov ecx, [ebp-0x14]
    L0032: call dword [0x13f2db84]
    L0038: fstp qword [ebp-0x10]
    L003b: fld qword [ebp-0x10]
    L003e: fld qword [ebp+0x8]
    L0041: fdiv dword [0x13fd2994]
    L0047: fmulp st1, st0
    L0049: fadd qword [ebp-0x8]
    L004c: sub esp, 0x8
    L004f: fstp qword [esp]
    L0052: mov ecx, [ebp-0x18]
    L0055: call dword [0x13f2db90]
    L005b: nop
    L005c: nop
    L005d: mov esp, ebp
    L005f: pop ebp
    L0060: ret 0x8
```

I played around a little with the Sample, and it really seems like every usage of the Salary property is inlined before the definition but not afterwards.

I have no idea what causes this, I guess the JIT compiler experts over at [coreclr](https://github.com/dotnet/coreclr) can answer this?

**Comment by mattwarren** [comment]
I don't think it's the same issue, but it does seem a little bit similar to the JIT bug that is [explained in this post](https://alexandrnikitin.github.io/blog/dotnet-generics-under-the-hood/#the-dessert)

The linked GitHub issue is an interesting read, especially the [very detailed answer from on of the JIT developers](https://github.com/dotnet/coreclr/issues/55#issuecomment-89026823)

**Comment by sonnemaf** [comment]
Must I (or you) report this problem to Microsoft? Where?

@mattwarren I'm not sure if this is related to the bug you mentioned in your reply. I'm not using Generics. But it could be. 

I just noticed that the problem also exists in the demo code you have on your GitHub page. If you move your Main() method below the Bar() and Foo() methods the Foo call is not inlined any more. I think this is quite serious.

```
    class Program {

        private static int Bar(string txt) {
            try {
                return int.Parse(txt);
            } catch {
                return 0;
            }
        }

        private static int Foo(string txt) {
            return int.Parse(txt);
        }

        static void Main(string[] args) {
            int a = Foo("12");
            int b = Bar("12");
        }
    }

```

**Comment by benaadams** [comment]
Raised issue https://github.com/dotnet/coreclr/issues/14615

**Comment by benaadams** [comment]
Looks like its an artifact of force Jitting rather than what would happen at runtime https://github.com/dotnet/coreclr/issues/14615#issuecomment-339097951

**Comment by szehetner** [comment]
The just released version 0.6.3 contains a proper fix for this. I implemented the suggestion by @AndyAyersMS in https://github.com/dotnet/coreclr/issues/14615 to build a call graph first and then do the jit compilation based on the calltree order (callers before callees). 

The sample above now shows all calls as inlined for me. Could you upgrade to the latest version and confirm if it works for you as well?

**Comment by sonnemaf** [comment]
It works! Great.
